### ISSUE TITLE
Text extraction

### ISSUE THREAD

**Issue by stephenfriedman** [pdfcpu/pdfcpu#122]
Any chance there has been thought given to a feature to return as a string the text inside a PDF? It seems that the `extract` command can yield content in the form of text files but this is still in the PDF encoding (https://pdfcpu.io/extract/extract_content.html). I would like the raw text. Seems like a valuable feature to be used with the extract command.

**Comment by hhrutter** [comment]
Hello!

Yes, that is work required for #4. I have been pushing back on this in order to focus on better font support in general first. I do hope we get there eventually.

**Comment by stephenfriedman** [comment]
Great thank you for the heads up!

**Comment by Yanwenjiepy** [comment]
Yes, I also hope that I can directly get the text content of the string type, so I will continue to pay attention.
I found a library that can directly get text content of type string, https://github.com/unidoc/unipdf-examples/blob/v3/text/pdf_extract_text.go.

**Comment by stephenfriedman** [comment]
@Yanwenjiepy I went with: https://github.com/google/go-tika

**Comment by nikolayk812** [comment]
So it is still not supported?

**Comment by stephenfriedman** [comment]
@nikolayk812 It appears that way

**Comment by fxtentacle** [comment]
PDF text extraction is quite difficult, because one first has to kind of render the PDF to get the positions of all the glyphs and then attempt to reconstruct the text from those glyphs. So I wouldn't fault the otherwise excellent pdfcpu project for not supporting it yet.

After reading the two other suggestions here, I decided against go-tika because I don't want to be dependent on having Java available on every computer running my Go app. As for unidoc, it used to be AGPL until v3.10.0 (released August 8, 2020) and since then they've made it a purely commercial library and the price of $1500 per developer seems a bit outlandish for a hobby project. So I'll go with the "old" AGPL version of Unidoc.

In case anyone else needs it, here's the link to their AGPL version:
https://github.com/unidoc/unipdf/tree/v3.9.0
which still contains source for the text extractor:
https://github.com/unidoc/unipdf/blob/v3.9.0/extractor/text.go
In contrast, the master branch is now overwritten with an encrypted mess:
https://github.com/unidoc/unipdf/blob/master/extractor/extractor.go

**Comment by stephenfriedman** [comment]
no fault given

**Comment by AlmogBaku** [comment]
please reopen

**Comment by wbdb** [comment]
Are there any plans as to when this functionality would be expected? I would greatly appreciate it as many other GO PDF readers do not have PDF Encryption v4 support. There seems to be no real alternative to UniPDF for GO at this time.

**Comment by pgmorgan** [comment]
I second @wbdb's point.  There is a dearth of free and open source libraries for Go to handle PDF's.  Please consider this feature.

**Comment by hhrutter** [comment]
Sorry folks, right now the focus is going Beta.
See #235

**Comment by profcoconut** [comment]
> PDF text extraction is quite difficult, because one first has to kind of render the PDF to get the positions of all the glyphs and then attempt to reconstruct the text from those glyphs. So I wouldn't fault the otherwise excellent pdfcpu project for not supporting it yet.
> 
> After reading the two other suggestions here, I decided against go-tika because I don't want to be dependent on having Java available on every computer running my Go app. As for unidoc, it used to be AGPL until v3.10.0 (released August 8, 2020) and since then they've made it a purely commercial library and the price of $1500 per developer seems a bit outlandish for a hobby project. So I'll go with the "old" AGPL version of Unidoc.
> 
> In case anyone else needs it, here's the link to their AGPL version: https://github.com/unidoc/unipdf/tree/v3.9.0 which still contains source for the text extractor: https://github.com/unidoc/unipdf/blob/v3.9.0/extractor/text.go In contrast, the master branch is now overwritten with an encrypted mess: https://github.com/unidoc/unipdf/blob/master/extractor/extractor.go

It seems this does not work any more. Text are truncated. 

```
It should be noted that fan-out, where a transaction depends on several transactions, and those
transactions depend on many more, is not a problem ... [Truncated - Unlicensed UniDoc - Get a license on https://unidoc.io]
Unlicensed copy of unidoc
To get rid of the watermark and keep entire text - Please get a license on https://unidoc.io
Page 6: 10. Privacy
```

**Comment by readonly24** [comment]
Hello.

I can do all my PDF processing with pdfcpu, except for the text extraction (to search for a string) I have to use [Apache PDFBox](https://pdfbox.apache.org/). It would be really 😎, if pdfcpu could handle this task as well.

**Comment by wbdb** [comment]
Another way :) https://github.com/klippa-app/go-pdfium . With 2 MB smaller than 12 MB PDF box

**Comment by huzpsb** [comment]
Hi hhrutter! Are you still interested in making this a feature after 6 years?  
I'm either going to create a PR or fork this, but if you'd like to delay it for now, please let me know so I can make more aggressive changes to the code.
Thanks!

**Comment by hhrutter** [comment]
A first cut of best-effort text extraction is planned to be part of one of the next releases possibly by July.

**Comment by hhrutter** [comment]
Hello!

I need to walk back the tentative July target. 
After diving into some foundational prototyping, the finding is a mid-year release is too aggressive for a project that prioritizes strict ISO 32000 compliance.

Implementing reliable text extraction isn't a basic string dump. 
Real-world PDFs are frequently non-compliant or structurally broken. 
Building a resilient layer requires handling low-level state recovery within cross-reference streams, managing legacy font encoding anomalies, and resolving highly fragmented content token flows without causing regressions across existing optimization and validation features.

I am moving this work into an isolated background development cycle alongside the upcoming Beta milestones and it will ship when it's fully stable and comprehensively tested.

Thanks for your patience as we take the time to build this the right way.

**Comment by hhrutter** [comment]
We developers are all aware of the world we are living in right now.

After careful consideration I have removed a msg promoting a specific AI tool, because this is not the right place.

Thx for understanding!

**Comment by wbdb** [comment]
No problem - it’s possible that someone weren’t aware of this free option for open-source projects, and it would have been helpful here, especially for larger tasks :)
